_posts/2021-10-14-MicroProfile-Reactive-Messaging-2.0-in-WildFly-25.adoc

```diff
@@ -0,0 +1,406 @@
+* Finally we return the massaged `Message` containing our received `PageVisit` instance, which will forward it to the `kafka-visits` stream.
+
+Another thing going on in this example, is that we're using an injected `Emitter` to 'fork' the sending of the received data to an additional location. In `fromFilter()`, if the page `3.html` was visited, we will also send the received `PageVisit` via the injected `Emitter`. This in turn will send the `PageVisit` instance on the `special` stream indicated in its `@Channel` annotation.
+
+The `special()` method, annotated with `@Incoming(`special`) receives messages from the `special` stream (i.e. the ones sent via the `Emitter`).
+
+When running the application, and clicking on the `3` link, you should see output in the server logs. Additionally, every click on any link will show up in the Kafka consumer logs mentioned in the example https://github.com/kabir/blog-reactive-messaging-2.0/blob/master/README.md[README]. So, in addition to being able to easily send data from user-initiated code, `Emitter` is useful for 'forking' streams, so you can send data to more than one location. This functionality was not present in version 1 of the specification.
+
+To map the `kafka-visits` stream to a Kafka topic we do the configuration in https://github.com/kabir/blog-reactive-messaging-2.0/blob/master/app/src/main/webapp/META-INF/microprofile-config.properties[microprofile-config.properties]:
+[source, properties]
+----
+mp.messaging.connector.smallrye-kafka.bootstrap.servers=localhost:9092
+
+mp.messaging.outgoing.kafka-visits.connector=smallrye-kafka
+mp.messaging.outgoing.kafka-visits.topic=page-visits
+mp.messaging.outgoing.kafka-visits.value.serializer=org.wildfly.blog.reactive.messaging.common.PageVisitsSerializer
+----
+
+This points the mapping towards `localhost:9092` to connect to Kafka, maps the `kafka-visits` stream to the  `page-visits` kafka topic, and specifies https://github.com/kabir/blog-reactive-messaging-2.0/blob/master/common/src/main/java/org/wildfly/blog/reactive/messaging/common/PageVisitsSerializer.java[PageVisitsSerializer] to be used to serialize the `PageVisit` instances that we send to Kafka. The https://docs.wildfly.org/25/Admin_Guide.html#MicroProfile_Reactive_Messaging_SmallRye[WildFly documentation] contains more detailed information about this configuration.
+
+If you deploy the application into WildFly, as outlined in the example https://github.com/kabir/blog-reactive-messaging-2.0/blob/master/README.md#deploy-the-microprofile-reactive-messaging-application[README], and you performed the optional step of connecting a Kafka consumer, you should see the output similar to this in the Kafka consumer terminal as you click the links in the application hosted at http://localhost:8080/app/:
+
+----
+frank	127.0.0.1app
+emma	127.0.0.13.html
```

Comment:
not sure if that's the code or the blog, but there is a missing space here, this would read better as

```
emma	127.0.0.1 3.html
```